internal/rc/rc.go

```diff
@@ -0,0 +1,495 @@
+// Copyright (C) 2015 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package rc provides remote control of a Syncthing process via the REST API.
+package rc
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/syncthing/protocol"
+	"github.com/syncthing/syncthing/internal/config"
+)
+
+// We set the API to use
+const APIKey = "592A47BC-A7DF-4C2F-89E0-A80B3E5094EE"
+
+type Process struct {
+	// Set at initialization
+	addr string
+
+	// Set by eventLoop()
+	eventMut          sync.Mutex
+	id                protocol.DeviceID
+	folders           []string
+	startComplete     bool
+	startCompleteCond *sync.Cond
+	stop              bool
+	localVersion      map[string]map[string]int64 // Folder ID => Device ID => LocalVersion
+	done              map[string]bool             // Folder ID => 100%
+
+	cmd   *exec.Cmd
+	logfd *os.File
+}
+
+// NewProcess returns a new Process talking to Syncthing at the specified address.
+// Example: NewProcess("127.0.0.1:8082")
+func NewProcess(addr string) *Process {
+	p := &Process{
+		addr:         addr,
+		localVersion: make(map[string]map[string]int64),
+		done:         make(map[string]bool),
+	}
+	p.startCompleteCond = sync.NewCond(&p.eventMut)
+	return p
+}
+
+// LogTo creates the specified log file and ensures that stdout and stderr
+// from the Start()ed process is redirected there. Must be called before
+// Start().
+func (p *Process) LogTo(filename string) error {
+	if p.cmd != nil {
+		panic("logfd cannot be set with an existing cmd")
+	}
+
+	if p.logfd != nil {
+		p.logfd.Close()
+	}
+
+	fd, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	p.logfd = fd
+	return nil
+}
+
+// Start runs the specified Syncthing binary with the given arguments.
+// Syncthing should be configured to provide an API on the address given to
+// NewProcess. Event processing is started.
+func (p *Process) Start(bin string, args ...string) error {
+	cmd := exec.Command(bin, args...)
+	if p.logfd != nil {
+		cmd.Stdout = p.logfd
+		cmd.Stderr = p.logfd
+	}
+	cmd.Env = append(os.Environ(), "STNORESTART=1", "STGUIAPIKEY="+APIKey)
+
+	err := cmd.Start()
+	if err != nil {
+		return err
+	}
+
+	p.cmd = cmd
+	go p.eventLoop()
+
+	return nil
+}
+
+// AwaitStartup waits for the Syncthing process to start and perform initial
+// scans of all folders.
+func (p *Process) AwaitStartup() {
+	for {
+		p.eventMut.Lock()
+		for !p.startComplete {
+			p.startCompleteCond.Wait()
+		}
+		p.eventMut.Unlock()
```

Comment:
No, should be fine I think. The `Cond` unlocks the mutex while it's waiting for a signal/broadcast.